opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/DefaultSessionUtils.java

```diff
@@ -60,4 +84,60 @@
         }
     }
 
+    /**
+     * NOTE: This is mostly copied from org.springframework.transaction.support.TransactionTemplate#execute
+     * but adds the ability to specify our own transaction definition.
+     */
+    private <V> V executeWithTransactionDefinition(TransactionDefinition transactionDefinition, Supplier<V> supplier) {
+        final TransactionStatus status = transactionManager.getTransaction(transactionDefinition);
+        V result;
+        try {
+            result = supplier.get();
+        }
+        catch (RuntimeException ex) {
+            // Transactional code threw application exception -> rollback
+            rollbackOnException(status, ex);
+            throw ex;
+        }
+        catch (Error err) {
+            // Transactional code threw error -> rollback
+            rollbackOnException(status, err);
+            throw err;
+        }
+        catch (Throwable ex) {
+            // Transactional code threw unexpected exception -> rollback
+            rollbackOnException(status, ex);
+            throw new UndeclaredThrowableException(ex, "TransactionCallback threw undeclared checked exception");
+        }
+        this.transactionManager.commit(status);
+        return result;
+    }
+
+    /**
+     * NOTE: This is a copy of org.springframework.transaction.support.TransactionTemplate#rollbackOnException
+     *
+     * Perform a rollback, handling rollback exceptions properly.
+     * @param status object representing the transaction
+     * @param ex the thrown application exception or error
+     * @throws TransactionException in case of a rollback error
+     */
+    private void rollbackOnException(TransactionStatus status, Throwable ex) throws TransactionException {
+        LOG.debug("Initiating transaction rollback on application exception", ex);
+        try {
+            this.transactionManager.rollback(status);
+        }
+        catch (TransactionSystemException ex2) {
+            LOG.error("Application exception overridden by rollback exception", ex);
```

Comment:
I'm assuming you want to print the rollback exception here, should be `ex2`. Same for below and for `Error err`.

Reply:
Not necessarily, those are re-thrown, so the caller can catch/print those.